### PR TITLE
[Core] Report worker backlog in GCS heartbeat

### PR DESCRIPTION
**Pull request by wuisawesome**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/protobuf/gcs.proto [hunk 1]
```diff
@@ -267,6 +267,9 @@ message ResourceDemand {
   // The number of requests for which there is no node that is a superset of
   // the requested resource shape.
   uint64 num_infeasible_requests_queued = 3;
+  // The number of requests of this shape still queued in CoreWorkers that this
+  // raylet knows about.
+  int64 backlog_size = 4;
```
Comment (wuisawesome) [inline src/ray/protobuf/gcs.proto:272]: Open question: should we also include some kind of `total_backlog_size` in `HeartbeatBatchTableData`?

File: src/ray/raylet/scheduling_queue.cc [hunk 2]
```diff
@@ -69,7 +69,13 @@ bool TaskQueue::AppendTask(const TaskID &task_id, const Task &task) {
   task_map_[task_id] = list_iterator;
   // Resource bookkeeping
   total_resource_load_.AddResources(task.GetTaskSpecification().GetRequiredResources());
-  resource_load_by_shape_[task.GetTaskSpecification().GetSchedulingClass()]++;
+  const auto &scheduling_class = task.GetTaskSpecification().GetSchedulingClass();
+  resource_load_by_shape_[scheduling_class]++;
+
+  int64_t backlog_size = task.BacklogSize();
+  if (backlog_size >= 0) { // Poor man's version of RayConfig::instance().report_worker_backlog()
```
Comment (wuisawesome) [inline src/ray/raylet/scheduling_queue.cc:76]: I haven't seen any other calls to `RayConfig::instance()` on the critical path, so I'm assuming it's too expensive to do here.

**Comment by wuisawesome** [comment]
cc @AmeerHajAli @rkooo567 @stephanie-wang

File: src/ray/gcs/gcs_server/gcs_node_manager.cc [hunk 3]
```diff
@@ -102,6 +104,10 @@ void GcsNodeManager::NodeFailureDetector::SendBatchedHeartbeat() {
         aggregate_demand.set_num_infeasible_requests_queued(
             aggregate_demand.num_infeasible_requests_queued() +
             demand.num_infeasible_requests_queued());
+        if (report_worker_backlog) {
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_node_manager.cc:107]: Just use ` RayConfig::instance().report_worker_backlog();` directly?

File: src/ray/raylet/scheduling_queue.cc (hunk 2 above)
Comment (rkooo567) [reply src/ray/raylet/scheduling_queue.cc:76]: Yeah I think accessing global field is pretty expensive. Not sure if it matters in raylet though.

Comment (rkooo567) [inline src/ray/raylet/scheduling_queue.cc:76]: >0?

**Review by rkooo567 — COMMENTED** [review]
Only reviewed tiny details.

File: src/ray/raylet/scheduling_queue.cc (hunk 2 above)
Comment (wuisawesome) [reply src/ray/raylet/scheduling_queue.cc:76]: -1 in the stand-in value if we aren't reporting the backlog, but i guess we don't actually need to track anything if the backlog is exactly 0 either.

Comment (rkooo567) [reply src/ray/raylet/scheduling_queue.cc]: Oh yeah. That's what I meant. It was supposed to be `>0`, but it was auto-formatted by github lol

File: src/ray/raylet/scheduling_queue.cc [hunk 4]
```diff
@@ -88,6 +94,10 @@ bool TaskQueue::RemoveTask(const TaskID &task_id, std::vector<Task> *removed_tas
   if (resource_load_by_shape_[scheduling_class] == 0) {
     resource_load_by_shape_.erase(scheduling_class);
   }
+  request_backlog_by_shape_[scheduling_class] -= it->BacklogSize();
+  if (request_backlog_by_shape_[scheduling_class] <= 0) {
```
Comment (stephanie-wang) [inline src/ray/raylet/scheduling_queue.cc:98]: Can this value ever go negative?

**Review by stephanie-wang — APPROVED** [review]
Oops sorry for the late review (next time, you can assign me so I see it).

It seems kind of strange to attach the backlog size to the `Task` class, but I guess that's just a sign that we need to refactor that interface...

**Comment by wuisawesome** [comment]
> Oops sorry for the late review

Lol no worries, it was a draft up until a few hours ago. Feedback is much appreciated and prompt :)

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
//python/ray/tests:test_global_state                                     FAILED in 3 out of 3 in 75.0s
 and LINT

**Comment by ericl** [comment]
LINT:
```
++++command -v clang-format
++++/home/travis/build/ray-project/tmp.lZwFYHdgTP/ci/travis/check-git-clang-format-output.sh
Running clang-format against branch 8f73014291bf1ccacb058bdb7399d5da0e38d81e, with hash 8f73014291bf1ccacb058bdb7399d5da0e38d81e
clang-format failed:
diff --git a/src/ray/common/task/task.h b/src/ray/common/task/task.h
index 0f9291de9..1fe9e24fc 100644
--- a/src/ray/common/task/task.h
+++ b/src/ray/common/task/task.h
@@ -35,7 +35,7 @@ class Task {
   /// \param message The protobuf message.
   /// \param backlog_size The size of the task owner's backlog size for this
   ///  task's shape.
-  explicit Task(const r
```

**Comment by ericl** [comment]
Do you have clang-format installed when you run scripts/format.sh?

**Comment by wuisawesome** [comment]
> Do you have clang-format installed when you run scripts/format.sh?

Yes, but I'm getting the impression that it's not running....

**Comment by ericl** [comment]
Seems like it's timing out the Windows build on test_global_state
```
2020-10-02T02:39:25.2097427Z ::test_backlog_report ================================================================================
2020-10-02T02:39:25.2098050Z ==================== Test output for //python/ray/tests:test_global_state:
```

Can you disable this test for Windows?

**Comment by rkooo567** [comment]
@wuisawesome Move the test to med size test (I did for this PR as well). I think it takes > 60 seconds to finish that test in Windows.https://github.com/ray-project/ray/pull/11129 (or you can wait until this PR is merged. I will merge it tomorrow)

**Pull request merged**